utils/src/test/java/software/amazon/awssdk/utils/cache/lru/LruCacheTest.java

```diff
@@ -0,0 +1,372 @@
+        verify(simpleValueSupplier, times(1)).apply(simpleTestKeys.get(4));
+        verify(simpleValueSupplier, times(1)).apply(simpleTestKeys.get(5));
+    }
+
+    @Test
+    void when_multipleCallers_oneCallerBlocks_simple() throws InterruptedException {
+        ExecutorService executor = Executors.newCachedThreadPool();
+        try  {
+            LruCache<Key, Value> cache = simpleCache.get();
+
+            for (int i = 0; i < 1000; i++) {
+                int numEntry = ThreadLocalRandom.current().nextInt(MAX_SIMPLE_TEST_ENTRIES - 1);
+                executor.submit(() -> cache.get(simpleTestKeys.get(numEntry)));
+            }
+
+            executor.shutdown();
+            assertThat(executor.awaitTermination(10, TimeUnit.SECONDS)).isTrue();
+        } finally {
+            executor.shutdownNow();
+        }
+    }
+
+    @Test
+    void when_multipleCallers_oneCallerBlock_doesnt_Work() {
+
+        try (WaitingSupplier waitingSupplier = new WaitingSupplier()) {
+            LruCache<Key, Value> cachedSupplier = LruCache.builder(waitingSupplier)
+                                                          .maxSize(5)
+                                                          .build();
+
+            waitingSupplier.permits.release(5);
+            waitFor(performAsyncGets(cachedSupplier, 8));
+
+            // Make extra sure only 2 "gets" actually happened.
+            waitingSupplier.waitForGetsToHaveFinished(2);
+        }
+    }
+
+    private void assertCacheState(LruCache<Key, Value> cache, Value result, int size, int index) {
+        assertThat(cache.size()).isEqualTo(size);
+        assertThat(result).isNotNull();
+        assertThat(result).isEqualTo(simpleTestValues.get(index));
+    }
+
+    private void populateAndVerify(LruCache<Key, Value> cache, int numEntries) {
+        IntStream.range(0, numEntries).forEach(i -> {
+            Value result = cache.get(simpleTestKeys.get(i));
+            assertThat(cache.size()).isEqualTo(i + 1);
+            assertThat(result).isNotNull();
+            assertThat(result).isEqualTo(simpleTestValues.get(i));
+            verify(simpleValueSupplier).apply(simpleTestKeys.get(i));
+        });
+    }
+
+    /**
+     * Asynchronously perform a "get" on the provided supplier, returning the future that will be completed when the "get"
+     * finishes.
+     */
+    private Future<Value> performAsyncGet(LruCache<Key, Value> supplier, Key key) {
+        return executorService.submit(() -> supplier.get(key));
+    }
+
+    /**
+     * Asynchronously perform multiple "gets" on the provided supplier, returning the collection of futures to be completed when
+     * the "get" finishes.
+     */
+    private List<Future<?>> performAsyncGets(LruCache<Key, Value> supplier, int count) {
+        List<Future<?>> futures = new ArrayList<>();
+        for (int i = 0; i < count; ++i) {
+            futures.add(performAsyncGet(supplier, simpleTestKeys.get(count % 2)));
```

Comment:
The name of the test method hides the very important point that the parallel calls are only across 2 keys. Is there a way to make that clearer in the method name or by making it a parameter?